### PR TITLE
HttpTracer: Add upstream_cluster tag to the tracing span (#1419)

### LINKED ISSUE #1419: Envoy should add a Binary Annotation to the Zipkin Span that contains the request target service
Let's assume the following scenario:
**service-a** sends a request to **service-b**. Envoy is configured to resolve **service-b** to either s**ervice-b-1** or **service-b-2** using a weighted cluster configuration.

The Zipkin Span sent from **service-a**'s envoy doesn't contain any information on the exact cluster to which the request was sent; The only thing that can be used is the **host_header** binary annotation which will just say "service-b", because the header was set by service-a and not the envoy. 

My proposal is to add a new binary annotation, **upstream-service-cluster**, that will contain the cluster name of the service that will receive the request.

What do you think?

### PR DESCRIPTION
**Pull request by clguimanMSFT**
Fixes #1419
* If the upstreamHost is available, then the tracer will add a tag with the upstream cluster name.
* Updated existing tests to work as before
* Added new test to verify the new behavior

### CODE REVIEW

**Comment by mattklein123** [comment]
@RomanDzhabarov can you please review?

File: test/common/tracing/http_tracer_impl_test.cc [hunk 1]
```diff
@@ -266,6 +266,7 @@ TEST(HttpConnManFinalizerImpl, NullRequestHeaders) {
   EXPECT_CALL(request_info, bytesSent()).WillOnce(Return(11));
   Optional<uint32_t> response_code;
   EXPECT_CALL(request_info, responseCode()).WillRepeatedly(ReturnRef(response_code));
+  EXPECT_CALL(request_info, upstreamHost()).WillOnce(Return(nullptr));
```
Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: can you also check that there is Times(0) calls to *span, setTag("upstream_cluster"....) in this case.

File: test/common/tracing/http_tracer_impl_test.cc [hunk 2]
```diff
@@ -278,6 +279,33 @@ TEST(HttpConnManFinalizerImpl, NullRequestHeaders) {
...
+  NiceMock<Http::AccessLog::MockRequestInfo> request_info;
+  auto upstream_host = std::make_shared<NiceMock<Upstream::MockHostDescription>>();
+  NiceMock<Upstream::MockClusterInfo> cluster_info;
+  const std::string expected_cluster_name = "<my_upstream_cluster>";
+
+  EXPECT_CALL(request_info, bytesReceived()).WillOnce(Return(10));
+  EXPECT_CALL(request_info, bytesSent()).WillOnce(Return(11));
+  Optional<uint32_t> response_code;
+  EXPECT_CALL(request_info, responseCode()).WillRepeatedly(ReturnRef(response_code));
+  EXPECT_CALL(request_info, upstreamHost()).WillRepeatedly(Return(upstream_host));
```
Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: can you reuse https://github.com/lyft/envoy/blob/3b20b62e43e2216845d1e436026aff6a2c5bebb5/test/mocks/http/mocks.h#L62 and do not create `auto upstream_host = std::make_shared<NiceMock<Upstream::MockHostDescription>>();` at all?

File: test/common/tracing/http_tracer_impl_test.cc (hunk 1 above)
Comment (clguimanMSFT) [inline test/common/tracing/http_tracer_impl_test.cc]: fixed

File: test/common/tracing/http_tracer_impl_test.cc [hunk 3]
```diff
@@ -278,6 +279,33 @@ TEST(HttpConnManFinalizerImpl, NullRequestHeaders) {
   finalizer.finalize(*span);
 }
 
+TEST(HttpConnManFinalizerImpl, UpstreamClusterTagSet) {
+  std::unique_ptr<NiceMock<MockSpan>> span(new NiceMock<MockSpan>());
+  NiceMock<Http::AccessLog::MockRequestInfo> request_info;
+  auto upstream_host = std::make_shared<NiceMock<Upstream::MockHostDescription>>();
+  NiceMock<Upstream::MockClusterInfo> cluster_info;
```
Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: and cluster_info also will not be needed

**Comment by RomanDzhabarov** [comment]
LGTM, few nits.

File: test/common/tracing/http_tracer_impl_test.cc (hunk 2 above)
Comment (clguimanMSFT) [inline test/common/tracing/http_tracer_impl_test.cc]: oh, I wasn't aware. I'll change it

File: test/common/tracing/http_tracer_impl_test.cc [hunk 4]
```diff
@@ -266,12 +266,39 @@ TEST(HttpConnManFinalizerImpl, NullRequestHeaders) {
...
+TEST(HttpConnManFinalizerImpl, UpstreamClusterTagSet) {
+  std::unique_ptr<NiceMock<MockSpan>> span(new NiceMock<MockSpan>());
+  NiceMock<Http::AccessLog::MockRequestInfo> request_info;
+  const std::string expected_cluster_name = "<my_upstream_cluster>";
+
+  EXPECT_CALL(request_info, bytesReceived()).WillOnce(Return(10));
+  EXPECT_CALL(request_info, bytesSent()).WillOnce(Return(11));
+  Optional<uint32_t> response_code;
+  EXPECT_CALL(request_info, responseCode()).WillRepeatedly(ReturnRef(response_code));
+  EXPECT_CALL(request_info, upstreamHost()).WillRepeatedly(Return(request_info.host_));
```
Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: I think this is already by default? you might need just change *request_info.host_->cluster_ = "something"

Comment (clguimanMSFT) [inline test/common/tracing/http_tracer_impl_test.cc]: fixed

Comment (RomanDzhabarov) [inline test/common/tracing/http_tracer_impl_test.cc]: `EXPECT_CALL(request_info, upstreamHost()).WillRepeatedly(Return(request_info.host_));` you can simply say

`EXPECT_CALL(request_info, upstreamHost())`, the behavior to return request_info.host_ is already done in ctor.

**Comment by RomanDzhabarov** [comment]
Last small nit and good to go.

**Pull request merged**